pkg/signature/kms/cliplugin/signer.go

```diff
@@ -0,0 +1,124 @@
+	"os"
+	"os/exec"
+
+	"github.com/sigstore/sigstore/pkg/cryptoutils"
+	"github.com/sigstore/sigstore/pkg/signature/kms"
+	"github.com/sigstore/sigstore/pkg/signature/kms/cliplugin/common"
+)
+
+var (
+	ErrorExecutingPlugin    = errors.New("error executing plugin program")
+	ErrorResponseParseError = errors.New("parsing plugin response")
+	ErrorPluginReturnError  = errors.New("plugin returned error")
+	ErrorInputKeyResourceID = errors.New("parsing input key resource id")
+	ErrorUnsupportedMethod  = errors.New("unsupported methodArgs")
+)
+
+// PluginClient implements kms.SignerVerifier with calls to our plugin program.
+type PluginClient struct {
+	kms.SignerVerifier
+	executable  string
+	initOptions common.InitOptions
+}
+
+// newPluginClient creates a new PluginClient.
+func newPluginClient(executable string, initOptions *common.InitOptions) *PluginClient {
+	pluginClient := &PluginClient{
+		executable:  executable,
+		initOptions: *initOptions,
+	}
+	return pluginClient
+}
+
+// invokePlugin invokes the plugin program and parses its response.
+func (c PluginClient) invokePlugin(ctx context.Context, stdin io.Reader, methodArgs *common.MethodArgs) (*common.PluginResp, error) {
+	pluginArgs := &common.PluginArgs{
+		InitOptions: &c.initOptions,
+		MethodArgs:  methodArgs,
+	}
+	argsEnc, err := json.Marshal(pluginArgs)
+	if err != nil {
+		return nil, err
+	}
+	cmd := exec.CommandContext(ctx, c.executable, common.ProtocolVersion, string(argsEnc))
+	cmd.Stdin = stdin
+	cmd.Stderr = os.Stderr
+	// We won't look at the program's non-zero exit code, but we will respect any other
+	// error, and cases when exec.ExitError.ExitCode() is 0 or -1:
+	//   * (0) the program finished successfuly or
+	//   * (-1) there was some other problem not due to the program itself.
+	// The only debugging is to either parse the the returned error in stdout,
+	// or for the user to examine the sterr logs.
+	// See https://pkg.go.dev/os#ProcessState.ExitCode.
+	stdout, err := cmd.Output()
+	var exitError *exec.ExitError
+	if err != nil && (!errors.As(err, &exitError) || exitError.ExitCode() < 1) {
+		return nil, fmt.Errorf("%w: %w", ErrorExecutingPlugin, err)
+	}
+	var resp common.PluginResp
+	if unmarshallErr := json.Unmarshal(stdout, &resp); unmarshallErr != nil {
+		return nil, fmt.Errorf("%w: %w", ErrorResponseParseError, unmarshallErr)
+	}
+	if resp.ErrorMessage != "" {
+		return nil, fmt.Errorf("%w: %s", ErrorPluginReturnError, resp.ErrorMessage)
+	}
+	return &resp, nil
+}
+
+// TODO: Additonal methods to be implemented
+
+func (c PluginClient) DefaultAlgorithm() string {
+	args := &common.MethodArgs{
+		MethodName:       common.DefaultAlgorithmMethodName,
+		DefaultAlgorithm: &common.DefaultAlgorithmArgs{},
+	}
+	resp, err := c.invokePlugin(context.TODO(), nil, args)
+	if err != nil {
+		log.Fatal(err)
+	}
+	return resp.DefaultAlgorithm.DefaultAlgorithm
+}
+
+func (c PluginClient) CreateKey(ctx context.Context, algorithm string) (crypto.PublicKey, error) {
```

Comment:
It'd be helpful to document the purpose of `handler`. If I'm understanding correctly, `handler` is helper functions for implementers in Go? Is there a reason why we need the indirection with the additional functions in handler/methods.go instead of `Dispatch` calling `pluginClient.Method()`?

Reply:
@loosebazooka No, it's the other way. `PluginClient.CreateKey()` via `invokePlugin()` and `Dispatch()` will always call `handler.CreateKey()`. Yes, it will inject timeouts. Authors of other languages would only have to look at `handler` as a reference.

@haydentherapper We could put all the content of `handler.CreateKey()` into `handler.Dispatch()`, but then it seems like a lot of behavior for one function.